src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala

```diff
@@ -0,0 +1,245 @@
+  // else, create and set descriptor based on kind, return descriptor
+
+  // clean_up
+  // forall tensor in hashmap, destroy tensor according to its kind
+  
+  def get_descriptor(t: Backend.Sym, kind: String)(implicit __pos: SourceContext): TOP = cudnnTensor2Desc.get(t) match {
+    case Some((desc, _)) => desc
+    case None => 
+      val shape = tensor_shape(t, useOldMetadata = true)
+      // val n = shape(0)
+      // val h = shape(1)
+      // val w = shape(2)
+      // val c = shape(3)
+      val n::h::w::c::_ = shape
+      kind match {
+        case "tensor" =>
+          val desc = new CUDNN_TENSOR_DESCRIPTOR(NEW_STRUCT(manifest[CUDNN_TENSOR_DESCRIPTOR], "cudnnTensorDescriptor_t").x)
+          CUDNN_CREATE_TENSOR_DESCRIPTOR(desc)
+          CUDNN_SET_TENSOR_4D_DESCRIPTOR(desc, CUDNN_NHWC, CUDNN_FLOAT, INT(n), INT(c), INT(h), INT(w))
+          cudnnTensor2Desc += ((t, (desc, kind)))
+          desc
+        case "filter" =>
+          val desc = new CUDNN_FILTER_DESCRIPTOR(NEW_STRUCT(manifest[CUDNN_FILTER_DESCRIPTOR], "cudnnFilterDescriptor_t").x)
+          CUDNN_CREATE_FILTER_DESCRIPTOR(desc)
+          CUDNN_SET_FILTER_4D_DESCRIPTOR(desc, CUDNN_NHWC, CUDNN_FLOAT, INT(n), INT(c), INT(h), INT(w))
+          cudnnTensor2Desc += ((t, (desc, kind)))
+          desc
+        case _ => throw new Exception("Unknown kind of cudnn tensor descriptor")
+      }
+  }
+
+
+  override def transform(n: Node): Backend.Exp = n match {
+    // NHWC as default layout
+    case Node(s, "tensor_conv", Backend.Const(tt: TensorType)::Backend.Const(anno:Anno)::(left:Backend.Sym)::(right:Backend.Sym)::
+      Backend.Const(padding:Seq[Int])::Backend.Const(strides:Seq[Int])::Backend.Const(dilation:Seq[Int])::
+      Backend.Const(alpha:Float)::Backend.Const(beta:Float)::_, _) =>
+      implicit val pos = Adapter.oldSourceMap(s)
+      // these are default settings
+      // TODO: perhaps move to base?
+      val tensor_dim = 4                // input tensor and input kernel should both be 4d
+      val hyperparam_dim = 2            // degree of freedom of padding, strides, and dilation
+      val layout = CUDNN_NHWC           // default tensor layout is batch x width x height x channel
+      val datatype = CUDNN_FLOAT        // only consider float
+      val mode = CUDNN_CONVOLUTION      // only consider convolution mode
+      // val padding = Seq(1, 1)
+      // val strides = Seq(1, 1)
+      // val dilation = Seq(1, 1)
+      // var alpha = 1.0f
+      // var beta = 0.0f   
+
+      // inputs
+      val left_shape = tensor_shape(left, useOldMetadata = true)
+      val right_shape = tensor_shape(right, useOldMetadata = true)
+      val left_operand = get_operand(left, anno)
+      val right_operand = get_operand(right, anno)
+      val leftTensor = new TENSOR(left_operand, useOldMetadata = true)
+      val rightTensor = new TENSOR(right_operand, useOldMetadata = true)
+      val sourceTensor = new TENSOR(s, useOldMetadata = true)
```

Comment:
I think `s` is not the sourceTensor, but rather the result tensor?

Reply:
`sourceTensor` should be deleted, I used it before but figured out that was incorrect, but forgot to change `sourceTensor` in `CUDNN_CONV_FWD` into `output`, which is a freshly allocated gpu array.

Reply:
fixed in 33dc2e5